### PR TITLE
added magnetic field to general hubbard model

### PR DESCRIPTION
**Pull request by bryano**

### CODE REVIEW

**Comment by bryano** [comment]
Ready for review. Failing tests unrelated to this PR.

**Comment by babbush** [comment]
I restarted tests. The failing test is one I mention in #477 - its started to get kind of annoying.

File: src/openfermion/hamiltonians/_general_hubbard.py [hunk 1]
```diff
@@ -39,6 +39,8 @@ def tunneling_operator(i, j, coefficient):
     return (FermionOperator(((i, 1), (j, 0)), coefficient) + 
             FermionOperator(((j, 1), (i, 0)), coefficient.conjugate()))
 
+def field_operator(i, coefficient=1.):
```
Comment (kevinsung) [inline src/openfermion/hamiltonians/_general_hubbard.py]: Please rename to `magnetic_field_operator`, which is more specific.

**Comment by bryano** [comment]
Done

File: src/openfermion/hamiltonians/_general_hubbard.py [hunk 2]
```diff
@@ -39,6 +39,8 @@ def tunneling_operator(i, j, coefficient):
     return (FermionOperator(((i, 1), (j, 0)), coefficient) + 
             FermionOperator(((j, 1), (i, 0)), coefficient.conjugate()))
 
+def magnetic_field_operator(i, coefficient=1.):
+    return number_operator(i, coefficient) - number_operator(i + 1, coefficient)
```
Comment (kevinsung) [inline src/openfermion/hamiltonians/_general_hubbard.py]: Isn't this assuming that i is up and i+1 is down? That seems incompatible with the more general indexing schemes you're supporting.

Comment (kevinsung) [reply src/openfermion/hamiltonians/_general_hubbard.py]: Ok, but there's a good chance we'll ultimately change the ordering to `(spin_index, site_index, dof_index)`. This ordering results in lower depth (by a factor of 2) circuits for Bogoliubov transformations in which orbitals of different spin don't interact. Most of your code is independent of the ordering, right? So I think we might as well make this one independent too.

**Comment by bryano** [comment]
Done

File: src/openfermion/hamiltonians/_general_hubbard.py [hunk 3]
```diff
@@ -39,8 +39,8 @@ def tunneling_operator(i, j, coefficient):
     return (FermionOperator(((i, 1), (j, 0)), coefficient) + 
             FermionOperator(((j, 1), (i, 0)), coefficient.conjugate()))
 
-def magnetic_field_operator(i, coefficient=1.):
-    return number_operator(i, coefficient) - number_operator(i + 1, coefficient)
+def difference_operator(i, j, coefficient=1.):
+    return number_operator(i, coefficient) - number_operator(j, coefficient)
```
Comment (kevinsung) [inline src/openfermion/hamiltonians/_general_hubbard.py]: Rename this to `number_difference_operator`

**Pull request merged**